modules/basis/basis-datasource/src/main/java/com/sap/xsk/listener/DelegatingObjectFactory.java

```diff
@@ -29,64 +31,64 @@ public Object getObjectInstance(Object obj, Name name, Context nameCtx, Hashtabl
-      if (objectInstance != null) {
-        LOGGER.fine("Object instance is obtained from default factory.");
-        return objectInstance;
-      }
     }
     throw new NamingException("Object instance with name [" + name + "] could not be created.");
   }
 
   private Object getObjectInstanceFromFactory(Object obj, Name name, Context nameCtx, Hashtable<?, ?> environment,
       final String referenceType) throws NamingException {
+
     final List<ObjectFactory> factories = getObjectFactories(referenceType);
     for (ObjectFactory factory : factories) {
-      Object objectInstance = null;
-      try {
-        objectInstance = factory.getObjectInstance(obj, name, nameCtx, environment);
-      } catch (NameNotFoundException nne) {
-        LOGGER.fine("Will move to next factory as facroty [" + factory + "] could not retrieve object instance for name [" + name
-            + "]. NameNotFoundException was thrown. " + nne.getMessage());
-      } catch (Exception e) {
-        LOGGER.severe("Exception is thrown by factory [" + factory + "] during retrieving object instance. Exception is: " + e);
-        throw new NamingException("Cannot create resource  object instance due to exception in the object factory");
-      }
-      if (objectInstance != null) {
-        LOGGER.fine("Object instance with name [" + name + "]is created by factory [" + factory + "].");
-        return objectInstance;
+      var res = createObject(obj, name, nameCtx, environment, factory);
+      if (res != null) {
+        LOGGER.fine(format("Object created for name {0} with factory {1}", name, factory));
+        return res;
       }
     }
-    return null;
+    return getObjectInstanceFromDefaults(obj, name, nameCtx, environment);
   }
 
   private Object getObjectInstanceFromDefaults(Object obj, Name name, Context nameCtx, Hashtable<?, ?> environment) throws NamingException {
     final Reference reference = (Reference) obj;
     RefAddr removedFactory = removeDelegatingObjectFactoryFromReference(reference);
-    Object objectInstance = null;
+    var factory = createFactory(reference);
+
+    if (removedFactory != null) {
+      LOGGER.info("Adding removed factory");
+      reference.add(removedFactory);
+    }
 
+    return createObject(obj, name, nameCtx, environment, factory);
+  }
+
+  private Object createObject(Object obj, Name name, Context nameCtx, Hashtable<?, ?> environment, ObjectFactory factory)
+      throws NamingException {
+    Object objectInstance = null;
     try {
-      String factoryClassName = reference.getFactoryClassName();
-      if (factoryClassName != null) {
-        ObjectFactory objectFactory = (ObjectFactory) getClass().getClassLoader().loadClass(factoryClassName).newInstance();
-        objectInstance = objectFactory.getObjectInstance(reference, name, nameCtx, environment);
-      }
+      objectInstance = factory.getObjectInstance(obj, name, nameCtx, environment);
     } catch (NameNotFoundException nne) {
```

Comment:
Isnt it better to throw some kind of exception instead of returning `null`

Reply:
It is part of the logic not to throw an exception